src/huntsman/pocs/autoflats.py

```diff
@@ -0,0 +1,428 @@
+        self._update(counts, exptimes, time_now)
+
+        # Take darks if we are finished
+        if self.is_finished:
+            self._take_darks(headers=headers)  # Implicit safety checking
+
+    def _calculate_target_counts(self):
+        """ Get the target counts and tolerance for each camera.
+        """
+        for cam_name, camera in self.cameras.items():
+            try:
+                bit_depth = camera.bit_depth.to_value(u.bit)
+            except NotImplementedError:
+                self.logger.debug(f'No bit_depth property for {cam_name}. Using 16.')
+                bit_depth = 16
+
+            self._target_counts[cam_name] = int(self._target_scaling * 2 ** bit_depth)
+            self._counts_tolerance[cam_name] = int(self._scaling_tolerance * 2 ** bit_depth)
+
+            self.logger.debug(f"Target counts for {cam_name}: {self._target_counts[cam_name]}"
+                              f"±{self._counts_tolerance[cam_name]}.")
+
+    def _update(self, average_counts, exptimes, time_now):
+        """ Update the sequence data with the previous iteration.
+        Args:
+            average_counts (dict): The average counts.
+            exptimes (dict): The exposure times.
+            time_now (datetime.datetime): The time that the exposures were started.
+        """
+        for cam_name in self.cameras.keys():
+            self._times[cam_name].append(time_now)
+            self._exptimes[cam_name].append(exptimes[cam_name])
+            self._average_counts[cam_name].append(average_counts[cam_name])
+
+        # Increment the sequence index
+        self._seqidx += 1
+
+    def _get_next_exptimes(self):
+        """ Calculate the next exposure times for all the cameras.
+        Returns:
+            dict: The cam_name: exptime pairs.
+        """
+        if self._seqidx == 0:
+            return self._initial_exposure_times
+        elapsed_time = current_time() - self._times[-1]
+        next_exptimes = {}
+        for cam_name in self.cameras.keys():
+            next_exptimes[cam_name] = self._get_next_exptime(cam_name, elapsed_time)
+        return next_exptimes
+
+    def _get_next_exptime(self, camera_name, elapsed_time):
+        """Calculate the next exposure time for the flat fields, accounting for changes in sky
+        brightness.
+        Args:
+            camera_name (str): The name of the camera.
+            elapsed_time (astropy.Quantity): The time between the previous exposure and now.
+        Returns:
+            astropy.Quantity: The next exposure time.
+        """
+        # Get data for specific camera
+        previous_exptime = self._exptimes[camera_name][-1]
+        target_counts = self._target_counts[camera_name]
+        average_counts = self._previous_counts[camera_name][-1]
+
+        # Calculate next exptime
+        exptime = previous_exptime * (target_counts / average_counts)
+        sky_factor = 2.0 ** (elapsed_time / 180.0)
+        if self.past_midnight:
```

Comment:
don't see where `past_midnight` is set. I guess elsewhere?

Reply:
See note in slack - can this algorithm be modified to not do the midnight check, but instead detect if sky is falling or rising automatically and set the next exptime accordingly? This might mean it can be used for day or dome flats.